File: 03-spring-examples/apps/seda/05-webflux-netty-server-client-netty/src/main/java/com/nalepa/demo/controller/NettyWebfluxControllerWithDedicatedCpuPool.java
package com.nalepa.demo.controller;

import com.nalepa.demo.common.Constants;
import com.nalepa.demo.common.DummyLogger;
import com.nalepa.demo.common.Operations;
import com.nalepa.demo.common.SomeResponse;
import com.nalepa.demo.common.monitored.ExecutorsFactory;
import com.nalepa.demo.httpclient.HttpClientFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Scheduler;
import reactor.core.scheduler.Schedulers;
import tools.jackson.databind.ObjectMapper;

import java.time.Duration;

@RestController
public class NettyWebfluxControllerWithDedicatedCpuPool {

    private final ObjectMapper objectMapper;
    private final org.springframework.web.reactive.function.client.WebClient webClient;
    private final Scheduler afterWebClientScheduler;

    public NettyWebfluxControllerWithDedicatedCpuPool(
            ExecutorsFactory executorsFactory,
            HttpClientFactory.HttpClientFactoryBean httpClientFactory,
            ObjectMapper objectMapper
    ) {
        this.objectMapper = objectMapper;
        this.webClient = httpClientFactory.createWebClient();
        this.afterWebClientScheduler = Schedulers.fromExecutor(
                executorsFactory.create(
                        ExecutorsFactory.ThreadPoolConfig.builder()
                                .threadPoolName("cpu.pool")
                                .threadsSize(Runtime.getRuntime().availableProcessors())
                                // queue size can be high,
                                // because when CPU is bottleneck, it's better to wait in queue
                                // than process all requests at the same time
                                .taskQueueSize(400)
                                .build()
                )
        );
    }

    @GetMapping("/endpoint/scenario/dedicatedCpuPool/{index}/{mockDelaySeconds}/{cpuOperationDelaySeconds}")
    public Mono<ResponseEntity<SomeResponse>> dummyEndpoint(
            @PathVariable String index,
            @PathVariable long mockDelaySeconds,
            @PathVariable long cpuOperationDelaySeconds
    ) {
        DummyLogger.log(this, "Start endpoint for index: " + index);

        return Mono.just(index)
                // .map {  } // eq: deserialize requestBody, remember about validation if needed
                .flatMap(i -> getData(index, mockDelaySeconds))
                .publishOn(afterWebClientScheduler)
                .map(byteArray -> {
                    try {
                        SomeResponse someResponse = objectMapper.readValue(byteArray, SomeResponse.class);
                        Operations.heavyCpuCode(cpuOperationDelaySeconds);
                        return someResponse;
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                })
                .map(ResponseEntity::ok);
    }

    private Mono<byte[]> getData(String index, long mockDelaySeconds) {
        long startTime = System.nanoTime();

        return webClient
                .get()
                .uri("http://localhost:8082/mock/{index}/{mockDelaySeconds}", index, mockDelaySeconds)
                .header(Constants.DUMMY_INDEX, index)
                .retrieve()
                // by default webClient thread will do deserialization, switch it to another thread pool if needed with publishOn
                .bodyToMono(byte[].class)
                .doOnNext(bytes -> {
                    Duration duration = Duration.ofNanos(System.nanoTime() - startTime);
                    DummyLogger.log(this, "Index: " + index + ". Got response from webClient after: " + duration);
                });
    }
}

File: 03-spring-examples/apps/seda/05-webflux-netty-server-client-netty/src/main/kotlin/com/nalepa/demo/controller/NettyWebfluxControllerWithDedicatedCpuPool.kt
package com.nalepa.demo.controller

import com.nalepa.demo.common.DUMMY_INDEX
import com.nalepa.demo.common.DummyLogger
import com.nalepa.demo.common.Operations
import com.nalepa.demo.common.SomeResponse
import com.nalepa.demo.common.monitored.ExecutorsFactory
import com.nalepa.demo.httpclient.HttpClientFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import tools.jackson.databind.ObjectMapper
import java.time.Duration

@RestController
class NettyWebfluxControllerWithDedicatedCpuPool(
    private val executorsFactory: ExecutorsFactory,
    private val httpClientFactory: HttpClientFactory,
    private val objectMapper: ObjectMapper,
) {

    private val webClient = httpClientFactory.createWebClient()

    private val afterWebClientScheduler =
        Schedulers.fromExecutor(
            executorsFactory.create(
                "CPU Bound pool waiting time took:",
                "CPU.for.afterWebClient",
                threadsSize = Runtime.getRuntime().availableProcessors(),
                taskQueueSize = Runtime.getRuntime().availableProcessors(),
            )
        )

    @GetMapping("/endpoint/scenario/dedicatedCpuPool/{index}/{mockDelaySeconds}/{cpuOperationDelaySeconds}")
    fun dummyEndpoint(
        @PathVariable index: String,
        @PathVariable mockDelaySeconds: Long,
        @PathVariable cpuOperationDelaySeconds: Long,
    ): Mono<ResponseEntity<SomeResponse>> {

        DummyLogger.log(this, "Start endpoint for index: $index")

        return Mono.just(index)
            // .map {  } // eq: deserialize requestBody, remember about validation if needed
            .flatMap {
                getData(index, mockDelaySeconds)
            }
            .publishOn(afterWebClientScheduler)
            .map { byteArray ->
                val someResponse = objectMapper.readValue(byteArray, SomeResponse::class.java)
                Operations.heavyCpuCode(cpuOperationDelaySeconds)
                someResponse
            }
            .map { ResponseEntity.ok(it) }
    }

    private fun getData(index: String, mockDelaySeconds: Long): Mono<ByteArray> {
        val startTime = System.nanoTime()

        return webClient
            .get()
            .uri("http://localhost:8082/mock/{index}/{mockDelaySeconds}", index, mockDelaySeconds)
            .header(DUMMY_INDEX, index)
            .retrieve()
            // by default webClient thread will do deserialization, switch it to another thread pool if needed with publishOn
            .bodyToMono(ByteArray::class.java)
            .doOnNext {
                val duration = Duration.ofNanos(System.nanoTime() - startTime)
                DummyLogger.log(this, "Index: $index. Got response from webClient after: $duration")
            }
    }


}